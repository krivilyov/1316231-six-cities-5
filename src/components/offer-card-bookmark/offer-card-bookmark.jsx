import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {fetchBookmarkOffers, updateOfferBookmarkStatus} from "../../store/api-actions";
import {getIsAuthorizedStatus} from "../../store/selectors";
import {AppRoute, offerBookmarkTypeProperty} from "../../const";
import OfferBookmarkContent from "../offer-bookmark-content/offer-bookmark-content";

const OfferCardBookmark = (props) => {
  const {offerId, offerBookmarkStatus, onChangeBookmark, isAuthorizedStatus} = props;

  const handleBookmarkClick = (evt) => {
    evt.preventDefault();
    onChangeBookmark(offerId, !offerBookmarkStatus);
  };

  return isAuthorizedStatus ? (
    <button className={`place-card__bookmark-button${
      offerBookmarkStatus ? ` place-card__bookmark-button--active ` : ` `}button`}
    type="button"
    onClick={handleBookmarkClick}
    >
      <OfferBookmarkContent
        offerBookmarkStatus={offerBookmarkStatus}
        offerBookmarkTypeProperty={offerBookmarkTypeProperty.card}
      />
    </button>
  ) : (
    <Link to={AppRoute.LOGIN} className={`place-card__bookmark-button
        ${offerBookmarkStatus ? ` place-card__bookmark-button--active ` : ` `}button`}
    type="button"
    >
      <OfferBookmarkContent
        offerBookmarkStatus={offerBookmarkStatus}
        offerBookmarkTypeProperty={offerBookmarkTypeProperty.card}
      />
    </Link>
  );
};

OfferCardBookmark.propTypes = {
  offerId: PropTypes.string.isRequired,
  offerBookmarkStatus: PropTypes.bool.isRequired,
  onChangeBookmark: PropTypes.func.isRequired,
  isAuthorizedStatus: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  isAuthorizedStatus: getIsAuthorizedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onChangeBookmark(offerId, bookmarkStatus) {
    dispatch(updateOfferBookmarkStatus(offerId, bookmarkStatus));
    dispatch(fetchBookmarkOffers());
  },
});

export {OfferCardBookmark};
export default connect(mapStateToProps, mapDispatchToProps)(OfferCardBookmark);
