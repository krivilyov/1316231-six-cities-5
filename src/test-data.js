export const testInitialState = {
  COMMON: {
    activeCity: `Amsterdam`,
    offers: [],
    reviews: [],
    sortingType: `popular`,
    mouseOverOfferId: null,
    changedBookmarkOffer: {},
    relatedOffers: [],
    bookmarkOffers: [],
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    authInfo: {},
  },
};
