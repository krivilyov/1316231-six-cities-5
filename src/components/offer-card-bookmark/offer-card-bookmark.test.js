
import React from "react";
import renderer from "react-test-renderer";
import {OfferCardBookmark} from "./offer-card-bookmark";
import configureMockStore from "redux-mock-store";
import {testInitialState} from "../../test-data";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";

const mockStore = configureMockStore();
const store = mockStore(testInitialState);

describe(`OfferCardBookmark should render correctly`, () => {
  it(`OfferCardBookmark should render correctly when  isAuthorizedStatus={false} and offerBookmarkStatus={false}`, function () {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferCardBookmark
                offerId={`1`}
                isAuthorizedStatus={false}
                offerBookmarkStatus={false}
                onChangeBookmark={() => {}}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferCardBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={true}`, function () {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferCardBookmark
                offerId={`1`}
                isAuthorizedStatus={true}
                offerBookmarkStatus={true}
                onChangeBookmark={() => {}}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferCardBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={true}`, function () {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferCardBookmark
                offerId={`1`}
                isAuthorizedStatus={false}
                offerBookmarkStatus={true}
                onChangeBookmark={() => {}}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`OfferCardBookmark should render correctly when  isAuthorizedStatus={true} and offerBookmarkStatus={false}`, function () {
    const tree = renderer
      .create(
          <Provider store={store}>
            <MemoryRouter>
              <OfferCardBookmark
                offerId={`1`}
                isAuthorizedStatus={true}
                offerBookmarkStatus={false}
                onChangeBookmark={() => {}}
              />
            </MemoryRouter>
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
