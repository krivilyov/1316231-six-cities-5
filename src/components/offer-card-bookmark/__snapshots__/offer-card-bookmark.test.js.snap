// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferCardBookmark should render correctly OfferCardBookmark should render correctly when  isAuthorizedStatus={false} and offerBookmarkStatus={false} 1`] = `
<a
  className="place-card__bookmark-button
         button"
  href="/login"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height="19"
    width="18"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</a>
`;

exports[`OfferCardBookmark should render correctly OfferCardBookmark should render correctly when  isAuthorizedStatus={true} and offerBookmarkStatus={false} 1`] = `
<button
  className="place-card__bookmark-button button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height="19"
    width="18"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</button>
`;

exports[`OfferCardBookmark should render correctly OfferCardBookmark should render correctly when isAuthorizedStatus={false} and offerBookmarkStatus={true} 1`] = `
<a
  className="place-card__bookmark-button
         place-card__bookmark-button--active button"
  href="/login"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height="19"
    width="18"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</a>
`;

exports[`OfferCardBookmark should render correctly OfferCardBookmark should render correctly when isAuthorizedStatus={true} and offerBookmarkStatus={true} 1`] = `
<button
  className="place-card__bookmark-button place-card__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height="19"
    width="18"
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</button>
`;
